File: client/guestregistration/src/app/services/card.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Card } from '../classes/card'

let httpOptions = {
  headers: new HttpHeaders({
    'Content-Type': 'application/json'
  }),
  withCredentials: true
};

@Injectable()
export class CardService {

  constructor(private http: HttpClient) { }

  /**
   * sends request to server to get all cards
   */
  getCards() {
    return this.http.get('http://localhost:8080/api/cards', httpOptions);
  }

  /**
   * sends request to server to create new card
   * @param card card data
   */
  postCard(card: Card) {
    let data = {
      "cardNo": card.cardNo,
      "validFrom": card.validFrom,
      "validThru": card.validThru,
      "officeId": card.location,
      "cardType": card.cardType
    }
    return this.http.post('http://localhost:8080/api/cards', data, httpOptions);
  }

  /**
   * sends request to server to delete card with provided id
   * @param id id of the card that will be deleted
   */
  deleteCard(id: String) {
    return this.http.delete('http://localhost:8080/api/cards/' + id,  httpOptions);
  }

  /**
   * sends request to server to update card with provided id
   * @param card updated card data
   * @param id id of the car that will be updated
   */
  putCard(card: Card) {
    
    let data = {
      "id": card.id,
      "cardNo": card.cardNo,
      "validFrom": card.validFrom,
      "validThru": card.validThru,
      "officeId": card.location,
      "cardType": card.cardType,
      "cardAvailable": card.cardAvailable
    }
    return this.http.put('http://localhost:8080/api/cards', data, httpOptions);
  }

  /**
   * sends request to server to get a single card with provided id
   * @param id id of the card
   */
  getSingleCard(id: string){
    return this.http.get('http://localhost:8080/api/cards/'+id,  httpOptions);
  }

  /**
   * sends request to the server to with search phrase to receive all
   * cards that contain that phrase
   * @param text search phrase
   */
  findCards(text: string, type: string){
    return this.http.get('http://localhost:8080/api/cards/find?query='+text+'&type='+type,  httpOptions);
  }

  getTypes(){
    return this.http.get('http://localhost:8080/api/cards/types', httpOptions);
  }

  getAvailableCards(){
    return this.http.get('http://localhost:8080/api/cards/available',  httpOptions);
  }
}

File: server/guestregistration/src/main/java/com/exadel/guestregistration/services/CardService.java
package com.exadel.guestregistration.services;

import com.exadel.guestregistration.exceptions.NotFoundException;
import com.exadel.guestregistration.exceptions.ParameterException;
import com.exadel.guestregistration.models.Card;

import java.util.List;

public interface CardService {

    /**
     * returns all cards in the database
     * @return
     */
    public List<Card> getCards();

    /**
     *
     * @param id id of the card
     * @return card with the provided id
     */
    public Card getCard(String id) throws NotFoundException;

    /**
     * creates card with supplied data
     * @param card card data
     * @return response entity
     */
    public Card createCard(Card card) throws ParameterException;

    /**
     * deletes card with provided id
     * @param id
     * @return response entity
     */
    public void deleteCard(String id) throws NotFoundException;

    /**
     * updates card with the same id
     * @param id card id that user wants to update
     * @param card updated card data
     * @return response entity with updated card data
     */
    public Card updateCard(Card card) throws ParameterException, NotFoundException;

    /**
     * returns list of cards that contains provided phrase
     * @param text search phrase
     * @return list of cards
     */
    public List<Card> findCards(String text, String type);

    /**
     * returns all available types of the card
     * @return
     */
    public List<String> getAllTypes();

    List<Card> getAvailableCards();
}
